### PR TITLE
samtools ampliconclip

### PR DESCRIPTION
**Pull request by whitwham**
New tool to clip amplicons, defined in a BED file, from the 5' end of a read.

### CODE REVIEW

**Comment by whitwham** [comment]
Pull request made to open program up to review.  Still missing tests and man page.  Program may undergo rapid change depending on evolving requirements.

**Comment by tseemann** [comment]
I think this is a good addition for viral amplicon work.
But should it be made a bit more generic (in name at least?).

It does what `ivar trim` does: https://github.com/andersen-lab/ivar
Hopefully @gkarthik can comment.

Does it handle overlapping sets of primers?

**Comment by gkarthik** [comment]
`ivar` based on htslib so a lot of overlap in code base! 
Based on [L181](https://github.com/whitwham/samtools/blob/dc8c6847afe93953eaa7f899f142ff9f5cea28b3/bam_ampliconclip.c#L181), should be able to handle overlapping primers: just iterating through and trimming all specified primers.

**Comment by whitwham** [comment]
> But should it be made a bit more generic (in name at least?).

Listen, don't you start.  We've had internal discussions about the name and nobody is happy about any of them.  Here is a list, if you come up with a good alternative let me know. :weary:

clip
amplicon-clip
primer-clip
pcr-clip
pcr-primer-clip
5prime-clip
read-5prime-coord-clip
targetcut
aclip
ampliconclip
clipamplicon
primerclip
primer-clip
pcr-primer-clip
pcrclip
mrclippy
amplicon-clippy-mc-clip-face
:scissors: (tool-that-was-formerly-known-as-amplicon-clip)
clippcr
bedclip

File: bam_ampliconclip.c [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+static int load_bed_file_pairs(char *infile, int get_strand, bed_pair_list_t *pairs,
+                                       int64_t *longest) {
+    hFILE *fp;
+    int line_count = 0, ret;
+    int64_t left, right;
+    kstring_t line = KS_INITIALIZE;
+    *longest = 0;
+
+    if ((fp = hopen(infile, "r")) == NULL) {
+        fprintf(stderr, "[ampliconclip] error: unable to open file %s\n", infile);
```
Comment (jmarshall) [inline bam_ampliconclip.c:77]: Somewhat trivial, but: could use `print_error_errno()` here.

**Comment by tseemann** [comment]
@whitwham i'll assume your comment is in jest, and good to see Sanger in the SARS-CoV-2 world :)

Q, If someone used a PE read-stitcher (FLASH, bbmerge etc) will `klippanator` be able to trim the 5' and 3' end of that psuedo SE read?

**Comment by whitwham** [comment]
> @whitwham i'll assume your comment is in jest,

Very much so.  Though the list is the real one I made of all the suggestions.  I should have stuck with the original name _clip_, it was less to type.

> and good to see Sanger in the SARS-CoV-2 world :)

Thank you.
 
> Q, If someone used a PE read-stitcher (FLASH, bbmerge etc) will `optimus-primer-klippanator` be able to trim the 5' and 3' end of that psuedo SE read?

I don't really have any experience of the output of any read-stitchers and I don't have the time to look at it at the moment.   `*clip` just works at the 5' prime ends currently.

**Comment by tseemann** [comment]
It might be important to support 3' and 5' end.
The ARTIC-style amplicon protocols currently tile into 400bp amplicons.
One approach is to not shear these, and sequence directly with 2x250 PE.
A read stitcher is used to overlap these to create a ~400bp SE end read, which consensus quality in thge overlapped region.

**Comment by whitwham** [comment]
@tseemann thanks for the info.  I'll see if I can find (or make) some stitched data.  It does sound like it would be useful.

**Comment by daviesrob** [comment]
A clipping both ends option might be useful for pacbio CCS reads too.

**Comment by daviesrob** [comment]
Merged so out pipeline people can play with it.  More updates to follow...

**Pull request merged**